### ISSUE TITLE
[R] Possible memory leaks due to SWIG_Fail and SWIG_Error definitions

### ISSUE THREAD

**Issue by Alzathar** [swig/swig#910]
The symbols `SWIG_Error`defined in [rfragments.swg:1](https://github.com/swig/swig/blob/master/Lib/r/rfragments.swg#L1) and `SWIG_Fail` defined in [rfragments.swg:3](https://github.com/swig/swig/blob/master/Lib/r/rfragments.swg#L3) used in `SWIG_exception_fail` have for effect to exit directly without managing memory cleanup (e.g. `std::string` conversion).

The other bindings (like Python and Matlab) use a `goto` statement in `SWIG_Fail` that transfers to a cleanup section.

I propose to use the same mechanism for the R bindings. The following patch should resolve the problem. I also transformed the definition of `SWIG_Error`to print only an error instead of printing a warning and exit. If necessary, I can create a PR.

```
diff --git a/Lib/r/rfragments.swg b/Lib/r/rfragments.swg
index b89212b..477d070 100644
--- a/Lib/r/rfragments.swg
+++ b/Lib/r/rfragments.swg
@@ -1,6 +1,6 @@
-#define SWIG_Error(code, msg) Rf_warning(msg); return Rf_ScalarLogical(NA_LOGICAL)
+#define SWIG_Error(code, msg) Rf_error(msg)
 
-#define SWIG_fail return Rf_ScalarLogical(NA_LOGICAL)
+#define SWIG_fail goto fail
 
 /* for raw pointers */
 #define SWIG_ConvertPtr(oc, ptr, ty, flags)             SWIG_R_ConvertPtr(oc, ptr, 
diff --git a/Source/Modules/r.cxx b/Source/Modules/r.cxx
index 301b49f..c401b2f 100644
--- a/Source/Modules/r.cxx
+++ b/Source/Modules/r.cxx
@@ -2086,8 +2086,10 @@ int R::functionWrapper(Node *n) {
   }
 
   /* Output cleanup code */
-  Printv(f->code, cleanup, NIL);
-  Delete(cleanup);
+  int need_cleanup = Len(cleanup) != 0;
+  if (need_cleanup) {
+    Printv(f->code, cleanup, NIL);
+  }
 
   /* Look to see if there is any newfree cleanup code */
   if (GetFlag(n, "feature:new")) {
@@ -2125,7 +2127,18 @@ int R::functionWrapper(Node *n) {
   if (destructor)
     Printv(f->code, "R_ClearExternalPtr(self);\n", NIL);
 
-  Printv(f->code, "return r_ans;\n}\n", NIL);
+  Printv(f->code, "return r_ans;\n", NIL);
+  
+  /* Error handling code */
+
+  Printv(f->code, "fail:", NIL);
+  if (need_cleanup) {
+    Printv(f->code, cleanup, NIL);
+  }
+  Printv(f->code, "  return R_NilValue;\n", NIL);
+  Delete(cleanup);
+  
+  Printv(f->code, "}\n", NIL);
   Printv(sfun->code, "\n}", NIL);
 
   /* Substitute the function name */
```

**Comment by ojwb** [comment]
Please open a PR (pull request) for your proposed patch - then it'll get automatically tested for most supported target languages on 3 platforms. If you just give us a patch file, we'd have to do all that testing by hand before we can be confident about merging.

It would also be good to include a testcase which checks that the memory clean-up has happened, if that's possible to do.

**Comment by Alzathar** [comment]
The PR is added in my TODO list!

Regarding the test case for the memory clean-up, I am not sure to know how to do it. In fact, I've just started to develop for R. Of course, tools like valgrind (more exactly memcheck) can be helpful but they cannot be integrated in unit tests.

Maybe someone with more experience with R can guide me on this.

**Comment by ojwb** [comment]
I know very little about R or SWIG's R backend, but if this clean up applies to a wrapped C++ class, then defining and wrapping a class like this which tracks how many live objects of the class exist would probably work:

```
class Foo {
    static unsigned count;
  public:
    Foo() { ++count; }
    ~Foo() { --count; }
    static unsigned get_count() { return count; }
};

static unsigned Foo::count = 0;
```

**Comment by Alzathar** [comment]
Finally, the PR #914 passed for the R language (see Travis job [2255.37](https://travis-ci.org/swig/swig/jobs/204221363)). The added test-case works too. 

However, I am not 100% satisfied by the final patch. The difficulty is on the way to manage the `goto` statements. Compared to other bindings, it seems that no proxy object is used for numerical types. Thus for some wrapped functions/methods, there is no use of the function `SWIG_exeception_fail` (and no `goto fail` call). In this case, the compiler warns or throws a compilation error. I forced the addition of this call but this is not elegant at all!

I do not have enough experience with this part of SWIG. Is there a way to know if the command `SWIG_exeception_fail` is called or not in the function code? Maybe by looking at the content of `f->code`? What do you think?

**Comment by ojwb** [comment]
There's no general way to know as the use of the label may be by code in a user-supplied typemap.  Inspecting the generated code is too complicated to do properly - e.g. a substring or regexp match is too error prone.

An unused label should only be a warning, but if you want to suppress that maybe annotate it with `SWIGUNUSED`, which will suppress such warnings for at least GCC and clang:

```
fail: SWIGUNUSED;
```

NB, the `;` after `SWIGUNUSED` is important!

Adding a dummy `goto` just seems to be playing warning whackamole - you're likely to get warnings about unreachable code or similar from the code you add to suppress the unused label warning.

**Comment by Alzathar** [comment]
Thanks @ojwb for your advice. This works like a charm with clang on MacOS. We have to wait for the CI verifications, but I think the PR is now complete and can be reviewed.

**Comment by wsfulton** [comment]
Thanks for the patches/fixes now committed, see #914.
